### DISCUSSION TITLE
Config Template jinja2 with ipaddr / netaddr

### DISCUSSION THREAD

**Discussion by routeriac** [netbox-community/netbox#12561] [Help Wanted!] [answered]
Hi,

I'm trying to do some IP calculations with config templates, but i'm not able to use ipaddr / netaddr.
This is just a plain simple example, I added an IP to an interface of the device in Netbox. The render config gives the exact variable as as output, but I wan't to replace the cidr notation by subnetmask, or strip the entire mask part for the usage of a router-id.

Is the jinja2 filter not usable by default? Should I add it to the JINJA2_FILTERS?
 

**Jinja2 Example to print the variable**
```
hostname {{ device.name }}
!
{% for interface in device.interfaces.all() %}
interface {{ interface }}
{% for ip in interface.ip_addresses.all() %}
  ip address {{ ip.address  }}
{% endfor %}
!
{% endfor %}
```

**Output Render Config**
```
hostname test_j2
!
interface GigabitEthernet0/0/0
  ip address 192.168.1.1/25
```

**Jinja2 Example to only print the IP**
```
hostname {{ device.name }}
!
{% for interface in device.interfaces.all() %}
interface {{ interface }}
{% for ip in interface.ip_addresses.all() %}
  ip address {{ ip.address | ipaddr('address') }}
{% endfor %}
!
{% endfor %}
```

**Output Render Config Error**
```
Traceback (most recent call last):
  File "/opt/netbox/netbox/dcim/views.py", line 2044, in get_extra_context
    rendered_config = config_template.render(context=context_data)
  File "/opt/netbox/netbox/extras/models/configs.py", line 251, in render
    template = environment.from_string(self.template_code)
  File "/opt/netbox/venv/lib/python3.8/site-packages/jinja2/environment.py", line 1105, in from_string
    return cls.from_code(self, self.compile(source), gs, None)
  File "/opt/netbox/venv/lib/python3.8/site-packages/jinja2/environment.py", line 768, in compile
    self.handle_exception(source=source_hint)
  File "/opt/netbox/venv/lib/python3.8/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<unknown>", line 6, in template
jinja2.exceptions.TemplateAssertionError: No filter named 'ipaddr'.
```

**Comment by kkthxbye-code** [comment, marked as answer]
The `ipaddr` filter seems to be an ansible thing. You can add it in JINJA2_FILTERS yeah.

In your specific case you should also just be able to do `{{ ip.address.ip }}`

**Reply by routeriac** [reply]
Thanks, this works. I will do some learning/testing with the default python ipaddress library to see if this works for me.
I like the netaddr/ipaddr because I can stack filters and do some more complex IP/subnet calculation.

**Reply by falz** [reply]
If you're able to add this filter please chime in with example, seems very useful for folks. Without it, netbox doesn't present the proper data required to actually render a config (ip without prefix, its netmask, its wildcard mask..)

**Reply by kkthxbye-code** [reply]
I think all your examples are available natively. ip.address is an IPNetwork object.

https://netaddr.readthedocs.io/en/latest/tutorial_01.html

Edit: Corrected link from ipaddress.IPNetwork to netaddr.IPNetwork

**Reply by falz** [reply]
Ah was not aware that was already ipnetwork obj, yep that should do it.

**Reply by markkuleinio** [reply]
I may miss some specific detail about this conversation but I don't think ip.address can be an `ipaddress.ip_network` object:

```
>>> ipaddress.ip_network("10.1.1.1/24")
Traceback (most recent call last):
...
ValueError: 10.1.1.1/24 has host bits set
```

but `ipaddress.ip_interface` would work:

```
>>> ipaddress.ip_interface("10.1.1.1/24")
IPv4Interface('10.1.1.1/24')
>>> ipaddress.ip_interface("10.1.1.1/24").ip
IPv4Address('10.1.1.1')
>>> ipaddress.ip_interface("10.1.1.1/24").netmask
IPv4Address('255.255.255.0')
>>> ipaddress.ip_interface("10.1.1.1/24").network.prefixlen
24
```

**Reply by kkthxbye-code** [reply]
Sorry, I linked the wrong docs. It's an netaddr.IPNetwork object.

```
{% for interface in device.interfaces.filter(ip_addresses__isnull=False) %}
  {% set ip_address = interface.ip_addresses.first() %}
  {{ ip_address.address.ip }}
  {{ ip_address.address.broadcast }}
  {{ ip_address.address.prefixlen }}
  {{ ip_address.address.netmask }}
{% endfor %}
```

<img width="287" alt="image" src="https://github.com/netbox-community/netbox/assets/400797/973cdaad-4dcc-4ace-bd53-8fd44c7f4f07">

This is all native, no casting to another object type.
